### PR TITLE
make symbol name demangling consistent

### LINKED ISSUE #717: Questionable location list for main.cold function
**Intention**

Trying to check that each of the location ranges match up with instruction boundaries and are within the function (or global).  Have example code at https://github.com/wcohen/quality_info/blob/master/dyninsttools/locrangechecks.C

**Describe the bug**

When running the locrangecheck tool on the binutils gold linker (ld-new) built on Fedora 30.  A number of the locations are marked as outside the current function main.cold.  For example it seems to be fetching the main function variable location lists for the main.cold function.  I verified that the function being inspected is main.cold, the function start address (f->addr())matches the start of main.cold.

**To Reproduce**

On Fedora30

wget https://kojipkgs.fedoraproject.org//packages/binutils/2.31.1/29.fc30/src/binutils-2.31.1-29.fc30.src.rpm
sudo dnf build-dep binutils-2.31.1-29.fc30.src.rpm
rpm -Uvh binutils-2.31.1-29.fc30.src.rpm
cd rpmbuild/SPEC
rpmbuild -ba binutils.spec
git clone https://github.com/wcohen/quality_info
cd quality_info/dyninsttools
make
./locrangechecks ~/rpmbuild/BUILD/binutils-2.31.1/gold/ld-new 

End up seeing:

# main.cold start:0x37410 
name=argc lowPC=3e540 hiPC=3e57c stClass=2 refClass=2
name = argc, k.lowPC=0x3e540 outside function
name = argc, k.hiPC=0x3e57c outside function
name = argc, k.lowPC=0x3e540 not valid boundary
name = argc, k.hiPC=0x3e57c not valid boundary
name=argc lowPC=3e57c hiPC=3e5a4 stClass=3 refClass=2
name = argc, k.lowPC=0x3e57c outside function
name = argc, k.hiPC=0x3e5a4 outside function
name = argc, k.lowPC=0x3e57c not valid boundary
name = argc, k.hiPC=0x3e5a4 not valid boundary
name=argc lowPC=3e57c hiPC=3e5a4 stClass=3 refClass=2
...

**Expected behavior**
Expect to get location list for the specific function being requested.

**System (please complete the following information):**
 - What platform  x86_64
 - What commit? dyninst-10.1.0-4.fc30.x86_64 on fedora 30
 - Maybe libc version glibc-2.29-28.fc30.x86_64

**Additional context**

### LINKED ISSUE #872: Make pretty and typed symbol names consistent and correct
**Intention**
Make pretty and typed symbol names be consistent for names that are mangled vs those that are not, and for the typed and pretty names with regards to clone suffixes, version suffixes and stabs suffixes.  Also improve the generation of the pretty name as mangled names that are not named methods or member of a namespace or class are not correct such as `operator()`.

**Describe the bug**
On linux and freebsd, the code to demangle the mangled symbol name into the pretty and typed versions is inconsistent.  There is one function P_cplus_demangle that performs the demangling but other pre and post processing is done inconsistently in functions that use P_cplus_demangle. For instance removing version suffixes and stabs suffixes is done outside of P_cplus_demangle, and is inconsistently performed.

For symbols that are not mangled using the Itanium C++ ABI, the pretty and typed name are the same as the symbol name include version and stabs suffixes, while for C++ mangled names these suffixes are discarded.  They should be discarded in all cases for consistency.  

For C++ mangled names and function clone suffixes in the name (starts with a '.'), the typed name includes the suffix in the output, while the pretty name discards this information, so clones of the same base name map the same pretty name, but different typed names.  For non-mangled names, the clone suffix is included, so this should present in the pretty name also; propose appending the clone suffix to the end of the pretty name.

The typed name for a C++ mangled name is equivalent to `c++filt -i -p <name>`, while the pretty name is not describable outside of inspecting the dedemangle function code, it should be equivalent to `c++filt -i <name>` with any clone suffixes appended.

**To Reproduce**
Review the pretty and typed names from processing libstdc++, libc, and code that has a clone function symbol (dyninst/examples codeCoverage testcc binary with current version of gcc).

**Expected behavior**

1. Typed and Pretty names should also remove any version and stabs suffixes
2. Pretty names should include any function clone suffixes
3. The pretty and typed names should be identical to the output of `c++filt` with the appropriate options `-I` and `-i -p`
4. This should all be implemented in a single place in the code

**System (please complete the following information):**
 - What platform (linux and freebsd)
 - Fix is on the demangle branch

**Additional context**

### PR DESCRIPTION
**Pull request by kupsch**
Make symbol name demangling consistent between C and C++, between pretty and typed names.  Removes "native compiler" concept (old C++ mangling format used with pgcc).

Fixes #717 
Fixes #872

### CODE REVIEW

File: symtabAPI/src/parseStab.C [hunk 1]
```diff
@@ -220,17 +209,7 @@ std::string Dyninst::SymtabAPI::parseStabString(Module *mod, int linenum, char *
...
-   {
-      name = mangledname;
-   } 
-   else 
-   {
-      name = demangled;
-   }
+   std::string name = P_cplus_demangle( mangledname );
 
    if ( name[0] != '\0' && stabstr[cnt] != ':' )
```
Comment (hainest) [inline symtabAPI/src/parseStab.C:214]: `std::string` isn't guaranteed to be null-terminated. A better check is `!name.empty()`. I know you didn't write this, but we may as well clean it up while we're here.

File: common/CMakeLists.txt [hunk 2]
```diff
@@ -45,12 +45,19 @@ set (SRC_LIST
...
 if (PLATFORM MATCHES freebsd)
   set (SRC_LIST ${SRC_LIST} 
     src/timing-freebsd.C
     src/freebsdKludges.C
     src/addrtranslate-sysv.C
     src/addrtranslate-freebsd.C
+    src/symbolDemangleWithCache.C
+  )
+  set (C_SRC_LIST ${C_SRC_LIST}
+    src/symbolDemangle.c
```
Comment (hainest) [inline common/CMakeLists.txt:60]: Could you remind me why we had to put this in a C-lang file by itself instead of in symbolDemangleWithCache?

**Review by hainest — CHANGES_REQUESTED** [review]

File: common/CMakeLists.txt [hunk 3]
```diff
@@ -45,12 +45,19 @@ set (SRC_LIST
...
+
 if (PLATFORM MATCHES freebsd)
   set (SRC_LIST ${SRC_LIST} 
     src/timing-freebsd.C
     src/freebsdKludges.C
     src/addrtranslate-sysv.C
     src/addrtranslate-freebsd.C
+    src/symbolDemangleWithCache.C
+  )
+  set (C_SRC_LIST ${C_SRC_LIST}
```
[the remark this answers is not in the document]
Comment (kupsch) [reply common/CMakeLists.txt:59]: It needed to be a C file as libiberty.h when included in C++, redefines basename in an incompatible way than the definition of basename in string.h:

extern char *basename(const char*);  // libiberty.h
extern "C++" const char *basename(const char*);   //string.h

**Review by hainest — APPROVED** [review]

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR889&status=OK

I'm seeing regressions in test5_9 on zeroah and coriander. I'm not sure what happened on cayenne.

**Comment by kupsch** [comment]
Do you have time for teams call to help me understand the testsuite?

Thanks,
Jim

**Comment by hainest** [comment]
I'm on Teams now, if you have time.

Looking at the log, we are getting these asserts firing on zeroah

    test_driver: /home/tim/paradyn/DEBY5hpE/dyninst/src/parseAPI/src/Function.C:509:
    void Dyninst::ParseAPI::Function::set_retstatus(Dyninst::ParseAPI::FuncReturnStatus):
    Assertion `!(_rs == NORETURN && rs == RETURN)' failed.
    
    test_driver: /home/tim/paradyn/DEBY5hpE/dyninst/src/symtabAPI/src/emitElfStatic-aarch64.C:169: 
    void Dyninst::SymtabAPI::emitElfStatic::getExcludedSymbolNames(std::set<std::__cxx11::basic_string<char> >&):
    Assertion `!EMIT_STATIC_ASSERT' failed.

Oddly, I'm not seeing those error messages on coriander. I would say let's focus on zeroah first, then we'll see what happened on coriander.

**Comment by kupsch** [comment]
@hainest, the demangle branch now passes the testsuite on coriander. You can test the other platforms now.

**Comment by hainest** [comment]
@kupsch  Could you rebase this on master?

**Comment by kupsch** [comment]
@hainest branch is rebased to master.  Testing can begin.

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR889&since_date=2020-10-21

Looking good so far. I need to install libiberty on the rest of my test systems before I can proceed further. I will let you know what I find.

**Comment by hainest** [comment]
LGTM. You can merge when you are ready.

**Pull request merged**